KIPs/kip-113.md

```diff
@@ -0,0 +1,252 @@
+| Managing address change timing                   | O             | O                                      | X                 | O               | O                        | O                      |
+| Awareness of user contracts after address update | X             | △<br> Latest GovParam address required | O                 | X               | O                        | X                      |
+
+In terms of explicit contract logic upgrade, ChainConfig, Precompiled-contract, and Bytecode-injection are recommended
+
+##### Usability from service-chain
+
+| Methods                      | [1] HeaderGov  | [2] ContractGov           | [3] SpareContract              | [4] ChainConfig | [5] Precompiled-contract          | [6] Bytecode-injection |
+| ---------------------------- | -------------- | ------------------------- | ------------------------------ | --------------- | --------------------------------- | ---------------------- |
+| Binary modification required | O              | O                         | X<br>Patch AddressBook address | O               | X<br> Patch the hardcoded address | O                      |
+| Contract deployment required | X<br> Registry | X <br> Registry, GovParam | X <br> Registry, AddressBook   | X<br> Registry  | X<br> Registry                    | X<br> AddressBook      |
+
+In terms of usability from service-chain, HeaderGov, ContractGov, ChainConfig, and Bytecode-injection are recommended
+
+##### Others
+
+| Methods                                            | [1] HeaderGov                 | [2] ContractGov               | [3] SpareContract             | [4] ChainConfig      | [5] Precompiled-contract | [6] Bytecode-injection |
+| -------------------------------------------------- | ----------------------------- | ----------------------------- | ----------------------------- | -------------------- | ------------------------ | ---------------------- |
+| Immediate panic at hardfork from out-of-date nodes | △<br> At GovernanceData block | X<br> At the activation block | △<br> At SpareContract update | O                    | O                        | O                      |
+| Registry initial data setup before hardfork        | O                             | O                             | O                             | O                    | O                        | X                      |
+| Special address (like 0x0…400)                     | X<br> vanity address          | X<br> vanity address          | X<br> vanity address          | X<br> vanity address | O                        | O                      |
+
+In terms of usability from ChainConfig, Precompiled-contract, and Bytecode-injection are recommended
+
+Overall, we choose ChainConfig because
+
+- users are mainly Klaytn nodes, not contracts so the awareness of user contracts after address update is less important
```

Comment:
I don't get the point of this statement. Can you explain little bit more why Klaytn nodes are main users? Klaytn nodes vs smart contracts seems little bit weird. Are you considering protocol vs smart contract?

Reply:
I meant users in terms of Registry readers. Registry is mainly read by Klaytn nodes like GovParam or KIP103. I'll clarify the term "user"

Reply:
let's use the term "subscriber"

Reply:
@ian0371 What about using the term "Core protocol" instead of "Klaytn nodes"?

Reply:
fixed in 6853068